llap-common/src/java/org/apache/hadoop/hive/llap/metrics/ReadWriteLockMetrics.java

```diff
@@ -0,0 +1,428 @@
+    @Override
+    public int compare(MetricsSource o1, MetricsSource o2) {
+      if (o1 != null && o2 != null
+          && o1 instanceof LockMetricSource && o2 instanceof LockMetricSource) {
+        LockMetricSource lms1 = (LockMetricSource)o1;
+        LockMetricSource lms2 = (LockMetricSource)o2;
+
+        long totalMs1 = (lms1.readLockWaitTimeTotal.value() / 1000000L)
+                        + (lms1.writeLockWaitTimeTotal.value() / 1000000L);
+        long totalMs2 = (lms2.readLockWaitTimeTotal.value() / 1000000L) 
+                        + (lms2.writeLockWaitTimeTotal.value() / 1000000L);
+
+        // sort descending by total lock time
+        if (totalMs1 < totalMs2) {
+          return 1;
+        }
+
+        if (totalMs1 > totalMs2) {
+          return -1;
+        }
+
+        // sort by label (ascending) if lock time is the same
+        return lms1.lockLabel.compareTo(lms2.lockLabel);
+      }
+
+      return 0;
+    }
+  }
+
+  /**
+   * Wraps a <code>ReadWriteLock</code> into a monitored lock if required by
+   * configuration. This helper is checking the <code>
+   * hive.llap.lockmetrics.collect</code> configuration option and wraps the
+   * passed in <code>ReadWriteLock</code> into a monitoring container if the
+   * option is set to <code>true</code>. Otherwise, the original (passed in)
+   * lock instance is returned unmodified.
+   *
+   * @param conf Configuration instance to check for LLAP conf options
+   * @param lock The <code>ReadWriteLock</code> to wrap for monitoring
+   * @param metrics The target container for locking metrics
+   * @see #createLockMetricsSource
+   */
+  public static ReadWriteLock wrap(Configuration conf, ReadWriteLock lock, 
+                                   MetricsSource metrics) {
+    Preconditions.checkNotNull(lock, "Caller has to provide valid input lock");
+    boolean needsWrap = false;
+
+    if (null != conf) {
+      needsWrap = 
+        HiveConf.getBoolVar(conf, HiveConf.ConfVars.LLAP_COLLECT_LOCK_METRICS);
+    }
+
+    if (false == needsWrap) {
+      return lock;
+    }
+
+    Preconditions.checkNotNull(metrics, 
+        "Caller has to procide group specific metrics source");
+    return new ReadWriteLockMetrics(lock, metrics);
+  }
+
+  /**
+   * Factory method for new metric collections.
+   * You can create and use a single <code>MetricsSource</code> collection for
+   * multiple R/W locks. This makes sense if several locks belong to a single
+   * group and you're then interested in the accumulated values for the whole
+   * group, rather than the single lock instance. The passed in label is
+   * supposed to identify the group uniquely.
+   *
+   * @param label The group identifier for lock statistics
+   */
+  public static MetricsSource createLockMetricsSource(String label) {
+    Preconditions.checkNotNull(label);
+    Preconditions.checkArgument(!label.contains("\""), 
+        "Label can't contain quote (\")");
```

Comment:
i guess this is a sanity check for URL stuff, then maybe you want to include stuff like `" %$&+,/:;=?@<>#%"`

Reply:
No, this is just because the label is content of the JSON output where label="content", therefore the content can't have a quote (or I would need to add escaping of quotes). All other characters should be fine (even if they don't make much sense).